slsim/Sources/QuasarCatalog/number_density_quasar.py

```diff
@@ -0,0 +1,224 @@
+    :param random_redshift_values: Array of randomly generated redshift values.
+    :type random_redshift_values: numpy.ndarray
+
+    Returns:
+    :rtype: list
+    :return: List of legend handles for each CDF plot.
+    """
+    # Compute the CDF data
+    cdf_data = compute_cdf_data(M_values, random_redshift_values)
+
+    # Create an empty list to collect artists for the legend
+    legend_handles_cdf = []
+
+    # Plot the CDF vs. M curve for each randomly generated redshift
+    for random_redshift, (sorted_M_values, cumulative_prob_norm) in cdf_data.items():
+        (line,) = plt.plot(
+            sorted_M_values, cumulative_prob_norm, label=f"z={random_redshift}"
+        )
+        legend_handles_cdf.append(line)
+
+    return legend_handles_cdf
+
+
+def inverse_cdf_fits_for_redshifts(M_values, random_redshift_values):
+    """Creates inverse Cumulative Distribution Function (CDF) fits for each randomly
+    generated redshift.
+
+    Parameters:
+    :param M_values: Array of absolute i-band magnitudes.
+    :type M_values: numpy.ndarray
+    :param random_redshift_values: Array of randomly generated redshift values.
+    :type random_redshift_values: numpy.ndarray
+
+    Returns:
+    :rtype: dict
+    :return: Dictionary of inverse CDF functions for each redshift.
+    """
+    # Compute the CDF data
+    cdf_data = compute_cdf_data(M_values, random_redshift_values)
+
+    inverse_cdf_dict = {}
+
+    for random_redshift, (sorted_M_values, cumulative_prob_norm) in cdf_data.items():
+        # Interpolate the inverse CDF function
+        inverse_cdf = interp1d(
+            cumulative_prob_norm,
+            sorted_M_values,
+            kind="linear",
+            fill_value="extrapolate",
+        )
+        inverse_cdf_dict[random_redshift] = inverse_cdf
+
+    return inverse_cdf_dict
+
+
+def generate_redshift_table(random_redshift_values, inverse_cdf_dict, seed=42):
+    """Generates random redshift values, associated inverse CDF values, and M values.
+
+    Parameters:
+    :param random_redshift_values: Array of randomly generated redshift values.
+    :type random_redshift_values: list or numpy.ndarray
+
+    :param inverse_cdf_dict: Dictionary of inverse CDF functions for each redshift.
+    :type inverse_cdf_dict: dict
+
+    :param seed: Seed for reproducibility (default: 42).
+    :type seed: int
+
+    Returns:
+    :rtype: astropy.table.Table
+    :return: Astropy table containing the generated data (Redshift, Inverse CDF Value, Associated M).
+    """
+    # Set the seed for reproducibility
+    np.random.seed(seed)
+
+    # Create an empty list to collect data for the table
+    table_data = {"Redshift": [], "Inverse_CDF_Value": [], "Associated_M": []}
+
+    # Generate random Inverse CDF value and associated M value for each redshift
+    for random_redshift in random_redshift_values:
+        # Find the inverse CDF function for the current random redshift
+        inverse_cdf = inverse_cdf_dict[random_redshift]
```

Comment:
Instead of using pre-computed inverse_cdf_dict, please call inverse_cdf_fits_for_redshifts() function here.